### DISCUSSION TITLE
Adding widget tooltips

### DISCUSSION THREAD

**Discussion by elParaguayo** [qtile/qtile#2251] [Ideas]
EDIT: For people joining the discussion now, this is the kind of thing we're discussing:
![slider](https://user-images.githubusercontent.com/946265/109401841-ef0aca00-7948-11eb-9333-37e882e9008b.gif)

I had made a "proof of concept" of a Tooltip mixin for widgets that would allow any widget to display a text tooltip (see #2249).

@m-col suggested tooltips could be even more useful if they could be widgets themselves (e.g. showing player controls for MPD, MPRIS2).

I think that's a great idea and would be very happy to have a go at implementing this. The point of this discussion is to discuss what's the best approach for doing it!

It would be preferable to use as much of the existing `_Widget` functionality as possible for these so maybe a new `_TooltipWidget` class could define the basic functionality needed.

The main question I have is how these are drawn. I don't think they can be configured as part of the `bar` because that's a window with a defined height and these widgets need to be drawn outside of that window. I think this means they need to be in a `Popup` window instead. In which case the `_TooltipWidget` should handle the creation of the window.

In addition, given that you can have more space in both the `x` and `y` dimensions, you could have some form of `TootipControl` object like buttons, sliders, text boxes, menus etc. that can be placed in the tooltip.

I think that's a pretty exciting idea but not straightforward!

Possible uses for this:

- Control media players, display artwork
- Display volume/brightness levels
- Display larger notifications inc. pictures/icons
- Show calendar under clock
- Menus (e.g. select wifi network)

Would welcome thoughts on this.

ps I'll also leave the Tooltip Mixin PR in draft for now as it may still have a use for some widgets that just want to show very basic information but, depending on the approach above, it may be possible to combine the two.

**Comment by elParaguayo** [comment]
Current thinking is that I'll have a main Tooltip object which accepts a number of "controls"/objects which can be placed on the tooltip window. I'll try to have a number of sizing/layout options for the controls (e.g. Tkinter-style grid, absolute positions). 

Don't think there's too much I can inherit from `_Widget` et al as I want a bit more flexibility on positioning and sizing.

**Reply by m-col** [reply]
Yeah that sounds like a good approach. I agree I don't think inheriting from `_Widget` makes much sense. In terms of configuration, I was initially picturing something like:
```py
date = widget.Clock(format='%e/%m/%g')
time = widget.Clock(tooltip=date)
```
Thus giving the "time" widget a tooltip that displays the "date". But as you mention, we'd want better control over configuring the tooltip's appearance... so perhaps something like:
```py
date = bar.Tooltip(widget.Clock(format='%e/%m/%g'), **config)
time = widget.Clock(tooltip=date)
```
where `**config` is any extra configuration options we want to give the tooltip. I think we could avoid lots of headaches if try to replicate the widget-positioning logic that `bar.Bar` uses to place widgets along itself.

**Reply by m-col** [reply]
`tooltip` would then be an option belonging to `_Widget`, so all widgets can receive a tooltip object and none need any special implementation within themselves to handle it.

**Reply by elParaguayo** [reply]
I agree, that would be a nice approach. I guess the question is, who's responsible for designing the tooltips? i.e. if it's the user, then they will just want to add a simple widget. However, if the "parent" widget defines the tooltip then it can be more complicated as the design is done behind the scenes.

My current approach is working on the framework to allow that latter option. For example, doing this for mpd or mpris2, we could have control buttons, artwork, track position, playlist etc. So you build the tooltip from all of these components and pass that to the widget internally rather than via a config. The config could then be used to hide it entirely if it wasn't wanted, or used to select different styles of (pre-built) tooltips etc.

For your example of the date, that was exactly the type of scenario I was imagining using the mixin for as it's perfect for providing additional text information. So something like: 
``` python
widget.Clock(format="%H:%M", tooltip_format="%d/%m/%y")
```

**Reply by elParaguayo** [reply]
Also conscious that the expanded functionality on a widget I refer to above is probably not a "tooltip"...

**Comment by elParaguayo** [comment]
Making some progress...

![musicwidget](https://user-images.githubusercontent.com/946265/109393430-7ccec100-7919-11eb-956b-6820c870e74d.gif)

**Reply by m-col** [reply]
Oh wow, that looks awesome! FWIW I don't think customisability is that important with something like this. It's probably much easier to do without, and having just one layout for the given widget.

**Reply by elParaguayo** [reply]
Yeah, you can have quite a lot of fun with this!

I've added a "slider" control (although there's no dragging) which means you can do things like this:

![slider](https://user-images.githubusercontent.com/946265/109401841-ef0aca00-7948-11eb-9333-37e882e9008b.gif)

I've also tried to simplify layout by recreating something like Tkinter's grid layout. For example, the music widget was done like this:

```python
controls = [
    TooltipImage(filename="~/play.svg",row=2, col=2, row_span=2, col_span=2),
    TooltipImage(filename="~/rw.svg",row=3, col=1),
    TooltipImage(filename="~/ff.svg",row=3, col=4),
    TooltipImage(filename="~/thylacine.jpeg",row=0, col=2, row_span=2, col_span=2),
    TooltipText(row=4,col_span=6),
    TooltipSlider(value=0.7, row=5, col_span=6)]

self.tooltip = TooltipGrid(rows=6, cols=6, width=200, height=200, parent=self, controls=controls)
```

Lots of issues to solve and lots of hard-coded values for now but I'll keep working on it.

**Reply by m-col** [reply]
I'm definitely impressed by this, but it looks like it will inevitably grow arms and legs so I wonder what is the best way to organise the code in terms of modules, or even repository. Then again I haven't seen all of the code so maybe it ain't too much :)

**Reply by elParaguayo** [reply]
Yeah. I was beginning to wonder that too. However, the code is actually quite light and, to be honest, should be much lighter as I've  just ended up copying lots from `bar` and `_Widget`. There's probably a way to refactor this but I'm focussing on functionality first (while trying not to get carried away!).

**Reply by elParaguayo** [reply]
Current code is here: https://github.com/elParaguayo/qtile/commit/b817a9785b4e42600aca3d052835df1b11eba70d

**Reply by DorianGray** [reply]
It would be wonderful if we could pass any widget to popup and have the widget control everything... like
```
widget = Image(filename='')
popup = Popup(widget=widget)
```

that widget could be displayed in multiple places or a part of multiple popups then.

Then if we could have something like
```
popup.attach(another_widget)
```

which would register the appropriate mouse callbacks. This is exactly why I've been working on mouse events slowly.

**Reply by elParaguayo** [reply]
Can you explain this a bit more? I'm not sure I'm entirely following your idea.

Do your first two lines of code represent setting up a "popup" with a single image and then using an `attach` method to connect it to a specific widget?

Are you envisaging users doing this in their configs or having this handled by the widgets?

I favour the latter and letting the user decide whether to enable the popup or not in their config. I suppose it would also be possible for users to create their own popup and attach it to widgets to replace the default.

As per some of my earlier comments, I've been trying to build a flexible framework for creating popups easily from smaller building blocks. Mouse already callbacks can be added to each of these blocks but could be abstracted into things like `on_{block_name}_click` which may help when building an `attach` method. There'd still need to be bespoke logic somewhere as popups would have different functionality/requirements.

It would certainly be possible to have some pre-configured popups (e.g. the music player one) and have this used by different widgets. If those widgets had standard methods like `cmd_play`, `cmd_next` etc. then I would think that we could build an `attach` method that connects those commands to the relevant buttons. So, if a user builds their own popup, they'd need to have the right control names.

**Reply by DorianGray** [reply]
I think there needs to be an easy way for normal users to add prebuilt widgets to their configs, but for widget developers it would be nice to have a more composeable approach so adding a tooltip doesn't change the type of the widget.

Like, a widget would just call attach on __init__ if it had a built-in popup but if a user wanted to add an onhover tooltip to, for instance, the clock widget, it would be nice for them to just be able to do that without creating a new class

**Reply by elParaguayo** [reply]
Ok. Let's take the mpris2 widget as an example.

I would want that widget to take a music popup (player controls, artwork etc). So, in the mpris2 widget code, we could include a reference to that popup. The user can decide whether to enable it or not with a config option. The user could also override the default popup with their own popup (they'd need to build this themselves but the framework should make this easy) and pass that via a config option too.

The mpris2 widget then needs some logic for attaching popups: i.e. check what controls are available, what text should be shown.

That feels like each widget needs custom `attach` code so, ideally, there'd be a method to handle this and widgets then just some sort of map that can be handled by the method so there's not too much custom code in widgets.

For mouse hover text tooltips, there is this the mixin that I was working on that started this whole discussion (#2249). It still feels like there's a place for that but maybe I'm wrong.

**Reply by DorianGray** [reply]
I've actually been trying to work on moving mouse events outside of widgets(ties to other work on using datastructures to improve mouse event performance) - where a widget would register/unregister it's events and placement on move/resize. That would make the model I'm describing make the most sense.

So, Widget for instance might implement mouse enter/leave whereas a descendant might add other mouse events simply by subscribing to them on a shared event machine.

**Reply by DorianGray** [reply]
Very explicitly...

```
from libqtile.mouse import hover

class Tooltip(Popup):
    def __init__(self, *args, **kwargs):
        hover.subscribe(area, self.on_hover)

    def on_hover(self, x, y):
        pass
```

**Reply by elParaguayo** [reply]
Ah. That might be getting beyond my skillset!

EDIT: We posted at the same time. Code looks interesting.

**Reply by DorianGray** [reply]
Attaching a popup in such an environment would be like...

```
class Popup(_Window):
    def attach(self, widget):
        mouse.hover.subscribe(widget.area, self.show)
```

**Reply by elParaguayo** [reply]
That would be nice because, as you say, it would enable me to bind actions for underlying controls at the popup level, without having to tie the two together.

How far have you got?

**Reply by DorianGray** [reply]
an r-tree would do really well as the backing datastructure for mouse events like the above...but if we decided to start keeping track of widget parents and children then we could simply walk the descendants like a tree to figure out where events should be fired. I kind of like the freeform version better because it allows for interesting things...like adding popups to decorate client windows on arbitrary areas.

**Reply by DorianGray** [reply]
not far, new baby 4 days ago so have been busy =) I've done all the research to implement it but haven't really started(just a few routines)

**Reply by elParaguayo** [reply]
Amazing. Congratulations.

**Reply by DorianGray** [reply]
Thanks. I haven't run these ideas past anyone really, except you and sort of @m-col . @tych0 @ramnes what do you guys think?

**Reply by tych0** [reply]
> @tych0 @ramnes what do you guys think?

I have to admit I don't really see the point of such things. The point of using a tiling window manager (at least in my mind) is that you can bind *everything* to a hotkey, so you never have to use the mouse at all, so having code that is triggered on hover mouse actions that offers buttons you can click doesn't make a lot of sense to me.

But there are lots of things Qtile has that I don't use, so if people want it it seems reasonable to include it (provided they come with tests, etc. so we can safely refactor things). It just means I likely won't be contributing anything.

**Reply by elParaguayo** [reply]
Fair enough! I did wonder if this was staying too much into desktop environment territory.

**Reply by tych0** [reply]
That's not to say that people shouldn't work on it if they want, or that the code shouldn't live in the qtile repo. It's just that I don't really understand the value of it personally. But I am strange :)

**Reply by elParaguayo** [reply]
I think that just means I haven't grasped the full power of tiling window managers!

**Comment by elParaguayo** [comment]
Question: What do we call these things? They're more than just "tooltips"?

**Reply by DorianGray** [reply]
Popups, lol. Yeah, who owns what logic could be really complicated here.

**Comment by elParaguayo** [comment]
While I think these changes could be quite exciting to implement, it is becoming a bigger job so I'd like to know if there is appetite for this.

In the meantime, I'll keep working on the basic framework as there are still some issues.

**Reply by tcld** [reply]
This seems like an incredibly cool feature! Too much customization, however, might also lead to less people using it. (Although...maybe that's wrong, considering we're talking about a tiling window manager.^^)

How much extra work does the customization entail on your end? Would it maybe be simpler to "just" give the user a widget that behaves a bit like a floating bar? It would then just be a very basic container for further widgets (horizontally or vertically) - and these are already sufficiently customizable to provide the user with whatever he wants (usually text or images with mouse callbacks).

**Reply by elParaguayo** [reply]
If I can find an easy way to bind commands/values then the customisation is really pretty straightforward (very easy for users to set almost every aspect of the popup).

I'm still trying to work out the right approach here.

**Comment by Barbaross93** [comment]
A cool idea for a possible tooltip for the groupbox widget: when hovering over a group icon/name on the widget, a tooltip with a preview of the group pops up. No idea how difficult that might be to implement though.

_Very_ cool stuff @elParaguayo, excited to see this develop!

**Comment by elParaguayo** [comment]
I haven't given up on this. Just been doing some other things.

**Comment by thingsiplay** [comment]
It's been a while, but I am still looking forward for official implementation of this. I know there is the [Popup Widget](https://qtile-extras.readthedocs.io/en/latest/manual/how_to/popup.html#extended-popups) from [Qtile extras](https://github.com/elParaguayo/qtile-extras), but would love to see this functionality builtin into all widgets themselves.

Is there any technical blocker why this is not official now?

**Reply by elParaguayo** [reply]
It's not going to be added to the main repo. There's too much extra code which isn't core functionality and it's not fair to ask the other devs to help maintain it.

**Comment by thingsiplay** [comment]
Fair enough. I was just curious about the current state. At least I know what it is and do not need to wait for, when implementing something myself.
